lib/Dialect/SPIRV/Serialization/Deserializer.cpp

```diff
@@ -722,6 +749,46 @@ LogicalResult Deserializer::processFunctionType(ArrayRef<uint32_t> operands) {
   return success();
 }
 
+LogicalResult Deserializer::processStructType(ArrayRef<uint32_t> operands) {
+  if (operands.size() < 2) {
+    return emitError(unknownLoc, "OpTypeStruct must have at least one member");
+  }
+
+  SmallVector<Type, 0> memberTypes;
+  for (uint32_t opNum = 1; opNum < operands.size(); ++opNum) {
+    Type memberType = getType(operands[opNum]);
+    if (!memberType) {
+      return emitError(unknownLoc, "OpTypeStruct references undefined <id> ")
+             << operands[opNum];
+    }
+    memberTypes.push_back(memberType);
+  }
+
+  // Check for layoutinfo
+  auto memberDecorationIterator = memberDecorationMap.find(operands[0]);
+  if (memberDecorationIterator != memberDecorationMap.end()) {
+    // Each member must have an offset
+    SmallVector<spirv::StructType::LayoutInfo, 0> layoutInfo;
+    const auto &offsetDecorationMap = memberDecorationIterator->second;
+    for (uint32_t memberNum = 0; memberNum < memberTypes.size(); ++memberNum) {
+      // Check that specific member has an offset
+      auto offsetIterator = offsetDecorationMap.find(memberNum);
+      if (offsetIterator != offsetDecorationMap.end()) {
+        layoutInfo.push_back(
+            static_cast<spirv::StructType::LayoutInfo>(offsetIterator->second));
```

Comment:
Prefer early exit, swap the the compare so that you can return early:

if (...) {
  return ...
}

Reply:
Thanks, I'll update it!

Reply:
If you change `memberDecorationMap` to `DenseMap<uint32_t, std::map<uint32_t, uint32_t>>`, you could use the same code as

```
SmallVector<spirv::StructType::LayoutInfo, 0> layoutInfo;
for (auto &memberDecoration : memberDecorationMap) {
   for (auto &memberIndexIterator : llvm::enumerate(memberDecoration)) {
      if (memberIndexIterator.index() != memberIndexIterator.value().second) {
            return emitError(...);
      }
      layoutInfo.push_back(static_cast<spirv::StructType::LayoutInfo>(memberIndexIterator.value().second));
   }
}
```
Does that not work?

Reply:
I think we prefer `llvm::DenseMap` instead of `std::map`. IIUC, `std::map` always does allocations so it's inferior to `llvm::DenseMap` here. Denis' current impl looks fine to me.

Reply:
Yes, it is a Nit. I am fine with it as well.

Reply:
@MaheshRavishankar I've missed that comment somehow, sorry about this, let me check it! Thanks!

Reply:
@MaheshRavishankar llvm::enumarate does not work for DenseMap, as far as I understand it:
```
‘class llvm::detail::enumerator_iter<llvm::detail::DenseMapPair<unsigned int, llvm::DenseMap<unsigned int, unsigned int, llvm::DenseMapInfo<unsigned int>, llvm::detail::DenseMapPair<unsigned int, unsigned int> > >&>’:
/home/denis/llvm-project/llvm/projects/mlir/lib/Dialect/SPIRV/Serialization/Deserializer.cpp:772:70:   required from here
/home/denis/llvm-project/llvm/include/llvm/ADT/STLExtras.h:52:40: error: no matching function for call to ‘begin(llvm::detail::DenseMapPair<unsigned int, llvm::DenseMap<unsigned int, unsigned int, llvm::DenseMapInfo<unsigned int>, llvm::detail::DenseMapPair<unsigned int, unsigned int> > >&)’
 using IterOfRange = decltype(std::begin(std::declval<RangeT &>()));
```
BTW, the order here is needed, because we processing OpTypeStruct %field0, ... %fieldN in order.
Thanks.